# Workgroup 5

# Group 6

Members:
1. Eljaer Eusebio
2. Claudia Vivas
3. Luis Sandoval
4. Andre Tapia

## Bootstrap

The bootstrap can be used to estimate the standard errors of the coefficients from a linear regression fit, or a
confidence interval for that coefficient. The power of the bootstrap lies in the fact that it can be easily applied to a wide range of statistical learning methods. 

In this replication, we use the Pennsylvania re-employment bonus experiment in order to compare treatment group 4 and the control group, for this purpose, we compute the standard errors of 1,000 bootstrap estimates for the $T4$, $female$ and $black$ variables.


In [73]:
#install.packages("ISLR2")
#install.packages("boot")

installing the source package 'ISLR2'

Warning message:
"package 'boot' is in use and will not be installed"

In [21]:
library(ISLR2)
set.seed(1)
library(boot)

Warning message:
"package 'boot' was built under R version 3.6.3"

### Loading the data

In [1]:
## loading the data
Penn <- as.data.frame(read.table("C:/Users/sniff/Documents/Estadistica aplicada/penn_jae.dat", header=T ))
n <- dim(Penn)[1]  
p_1 <- dim(Penn)[2] 
Penn<- subset(Penn, tg==4 | tg==0) 
attach(Penn)

In [4]:
T4<- (tg==4)
summary(T4)

   Mode   FALSE    TRUE 
logical    3354    1745 

In [5]:
head(Penn)

abdt  tg inuidur1 inuidur2 female black hispanic othrace dep q1 ... q5 q6
1  10824 0  18       18       0      0     0        0       2   0  ... 1  0 
4  10824 0   1        1       0      0     0        0       0   0  ... 1  0 
5  10747 0  27       27       0      0     0        0       0   0  ... 0  0 
12 10607 4   9        9       0      0     0        0       0   0  ... 0  0 
13 10831 0  27       27       0      0     0        0       1   0  ... 1  0 
14 10845 0  27       27       1      0     0        0       0   0  ... 1  0 
   recall agelt35 agegt54 durable nondurable lusd husd muld
1  0      0       0       0       0          0    1    0   
4  0      0       0       0       0          1    0    0   
5  0      0       0       0       0          1    0    0   
12 0      1       0       0       0          0    0    1   
13 0      0       1       1       0          1    0    0   
14 0      0       1       0       0          1    0    0

In [11]:
dim(Penn)

[1] 5099   23

## Example 1

First, we resample the data with replacement, and the size of the resample must be equal to the size of the original data set. Then the statistic of interest is computed from the resample from the first step. We repeat this routine many times to get a more precise estimate of the Bootstrap distribution of the statistic.

We create $boot.fn()$ which takes in the $Penn$ data set as well as a set of indices for the observations, and returns the intercept and slope estimates for the linear regression model. We then apply this function to the full set of  5099  observations in order to compute the estimates 16 variables.

Then, with $boot()$ function to compute the standard errors of 1,000 bootstrap estimates for the intercept and slope terms. Note that we are using $log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd$ formula for estimation, so bootstraping is goíng to estimate 16 variables (intercept included). 

In [8]:
boot.fn <- function(Penn, index)
  coef(lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data = Penn, subset = index))

In [9]:
lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data = Penn)


Call:
lm(formula = log(inuidur1) ~ T4 + (female + black + othrace + 
    factor(dep) + q2 + q3 + q4 + q5 + q6 + agelt35 + agegt54 + 
    durable + lusd + husd), data = Penn)

Coefficients:
 (Intercept)        T4TRUE        female         black       othrace  
    2.178462     -0.071692      0.126368     -0.293768     -0.472445  
factor(dep)1  factor(dep)2            q2            q3            q4  
    0.029867      0.096187      0.073678     -0.038507     -0.054949  
          q5            q6       agelt35       agegt54       durable  
   -0.144178      0.003361     -0.162772      0.229667      0.126557  
        lusd          husd  
   -0.175353     -0.105225  


In [12]:
boot.fn(Penn, 1:5099)

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.178462326 -0.071692484  0.126368328 -0.293767980 -0.472445058  0.029866899 
factor(dep)2           q2           q3           q4           q5           q6 
 0.096186517  0.073678072 -0.038506537 -0.054949195 -0.144177912  0.003361318 
     agelt35      agegt54      durable         lusd         husd 
-0.162772168  0.229666708  0.126557359 -0.175352572 -0.105224727

 The `boot.fn()` function can also be used in order to create
 bootstrap estimates for the intercept and slope terms by randomly sampling from among the observations with replacement. Here we give two examples.

In [13]:
sample(1:5099, 5099, replace = T)

[1] 2111 1266 1521 4741 2008 1153 3754 2053 4728 1350  327 1117 3174 3718
  [15]  534  752 3359  254 4709 2526  240  532 2109 4239 3569 4460 1598  885
  [29] 1697  766 4926 3193 5052 1072 2787  791 2666 3035  872 2080 2169  762
  [43] 2230 2174 2040  372 5092 1209 4799  313 2224 1765 3906 3904 2554 2917
  [57] 2280 3720 2994  815 1176 4205  915 3871 2064 2467 4464 4682 4152 3063
  [71] 2468 2309 3416  870 2266 3897 3608  849   12 3761  422 1974 2114 3389
  [85] 2899 3333 2690 4305 3170 3525 2661 4417 3853  273 1640 2001 3669 4155
  [99] 4167  920 3557 1281  737 3408 2665 2417 4036 2334  256  499 4760  116
 [113] 3596 1214 2558 1011 3010 4652 3192  382 3754 4817 2140 2845 3580 2839
 [127]  325  485 3009 4807 4510 4580  727 1809  278 2600 2220 3048 1997 2337
 [141] 4217 4878 2665 3986 1748 1929 2925 1524 2348 2107 4239  260 3033  957
 [155] 4000 2103 1544 2390 3380 3253 1124  441 4455 1812  621 1916  234  936
 [169] 2110 2754 3425 2598 1274    2 2875 4758 3185 1265 1415 3884 4288  731
 [183] 1651 1870 3984  687 1230 1871 3630 1411 1072 2090 4615 1017 4319 2671
 [197] 1772  974 4186  205 3289 1954 4365 1549 4572 3205 4718  725 1862 3792
 [211] 2644 4431 2293  176 2164  104 2708 4496 4421 2168 2359 3848 3823  979
 [225] 3342 3595 3044 4752 3053 1640  870 3536 1720  143 4188 1889 4431 3529
 [239] 1357 2721 4499 2999 1566 1327  109 2561 2292 4611 1995  347 1592  182
 [253] 4730  500 2400 3474 1705 3604 3784 1108 3746 2406 4939 1120 4832  365
 [267] 2662 2923 3437 3238 1749 4365 4377 1906 2812 1964 4919  351 1549 4952
 [281] 3716 4041   62 2981 3329 4070 3884 4730 4588 1974  282 2684 3371 1202
 [295]  515 3018  127 3317 1743 3831 1254 1274   71 5054 4588  460 2004  895
 [309]  746 5085  902  686 2075 1055 3060  726 5028 2765 3276 3054 2628 2999
 [323] 4240  244  646  534 1529 1784  251 1164 1165 3881 2342 1064 1730 1189
 [337]  503 2410 3136  569 1879 3390 4312 2670 2782 4676  764 2000 3973 2057
 [351] 5071 3260  611 3358 2520 2311 1843 3723 3421 1656 4077  779 1367 4361
 [365] 1501 1988 2117 3209 3727 2340 1889 4387 2355 2395 2361 4360  597 4816
 [379] 1099  131 2955 4584 4552  451 4491  874 1408 1282 4935 1491 4118 3746
 [393] 2756 1491  985 4981 1443  889 1572 1335 2220 2631 4338 1357 2043 4927
 [407] 1058 1158 4104 3895  219 4710 3254 2254 3262 3139 2806 3214   48    6
 [421]  483 3482 4791 1129 1711 3106 3532 1898   96 2182 1658 2893  689 4818
 [435] 1719  415 2057  551  915 1301 4989 1611 4735 1142 3103  184 1928 4659
 [449] 1049  482 2945 2297 1206 1577 4257 4470  660 1835 3315  124  345 4325
 [463] 2788 4742 3839 1637 2003 3675 1000 4881 4531 1399 2061  131 3252 4292
 [477] 1092  293 2486 3538 2175   42 2511 3286 2404 1132 2041 3274 1049 2069
 [491] 1600  839 2813  120 3607 2517  325 2869 1960 2461 1210 3765  121  550
 [505] 2445 1307 3305 3402 1952 2099 3613 1379 2485 2114 1040 4200 4104 4696
 [519] 1974 4427  653 5041 4619 3227 4650  534 2849 4643 1466 2698  571 2407
 [533] 2859 2547 3836 3360 1299 4371 2738 1814 2180 3378  577  458 4569  628
 [547] 3271 4051 2995 3612  425 4562 2243 4822 3314 1792 1912 4614 2697 2818
 [561] 1856 1288 1681 3318 2504 4183 2607 1231 4416 1744  120  252 4312 4576
 [575] 1895  675  891 3631  266 4394    2 1836 4100 2051  215 1173   40 2811
 [589]  447  174 4444 5078 3735 4005  415  838 4414  316 1917  453 1221 3772
 [603] 1403 3145  512 2979 4681 3407 3313 4984 2108 3010  621 2826 4453 3608
 [617] 3925 1568 3896 4823 4458  713 1000 2583 3828 1499 5038 1547 4429  831
 [631]   75 4211 4825 2306  967 2469 1691 2416 4968 4181 3608  833 1183  572
 [645] 3978  883 1689 3860  489  983 2505 2462 1438  778 1648 3345 3053 4463
 [659] 1558  454 1875 3299 2699 3023  604 1579 4947 1779 3478  506 3078 4858
 [673]   90 3042 1037 4876 3668   96 2350 3953 3512  699 1046 1538  887 3426
 [687] 3946 3242  100 1212  399 4994 4803 2391 3948 2305 4589 4111 4057 1847
 [701] 4294 2880  103 4877  369 3390 4447 4572 4379 2798 4444 3997 4987 4190
 [715] 3400 1465 2196 1005 4897 3823  796 3153 1037 1590  382 1740 3629 1461

In [17]:
set.seed(1)
boot.fn(Penn, sample(5099, 5099, replace = T))
boot.fn(Penn, sample(5099, 5099, replace = T))

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.258670234 -0.117228952  0.142462443 -0.374373286 -0.713971846  0.009594995 
factor(dep)2           q2           q3           q4           q5           q6 
 0.090434910  0.005622350 -0.133365893 -0.114376148 -0.180682473  0.036319276 
     agelt35      agegt54      durable         lusd         husd 
-0.153838318  0.185651787  0.085436134 -0.156716042 -0.092207747

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
  2.07020431  -0.07586211   0.14495743  -0.40061393  -0.68451940   0.08656088 
factor(dep)2           q2           q3           q4           q5           q6 
  0.17472375   0.15982631   0.08429811  -0.01441098  -0.02045042   0.04632322 
     agelt35      agegt54      durable         lusd         husd 
 -0.17008943   0.24289153   0.12353482  -0.13700666  -0.12442404

Next, we use the `boot()` function to compute the standard errors of 1,000 bootstrap estimates for the intercept and slope terms.

In [72]:
boot1 = boot(Penn, boot.fn, 1000)
boot1


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Penn, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
         original        bias    std. error
t1*   2.178462326  8.628683e-03  0.15210515
t2*  -0.071692484 -5.257804e-04  0.03461895
t3*   0.126368328 -1.393659e-03  0.03627517
t4*  -0.293767980 -1.068414e-03  0.05955501
t5*  -0.472445058 -5.911082e-03  0.23654083
t6*   0.029866899  1.093333e-03  0.05391334
t7*   0.096186517  2.541850e-03  0.04487310
t8*   0.073678072 -8.888514e-03  0.14906345
t9*  -0.038506537 -7.986062e-03  0.15196746
t10* -0.054949195 -8.523033e-03  0.14940522
t11* -0.144177912 -7.718904e-03  0.14897661
t12*  0.003361318 -5.889983e-03  0.16105233
t13* -0.162772168 -2.476379e-03  0.03573635
t14*  0.229666708  2.849669e-03  0.05711884
t15*  0.126557359 -6.017732e-04  0.04733377
t16* -0.175352572 -2.588328e-05  0.04022709
t17* -0.105224727  1.861795e-03  0.04518485

In [69]:
s =summary(
    lm(log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd)), data = Penn
    )$coef
s

Estimate     Std. Error t value     Pr(>|t|)    
(Intercept)   2.178462326 0.15901507 13.69972271 5.631260e-42
T4TRUE       -0.071692484 0.03546326 -2.02159887 4.327012e-02
female        0.126368328 0.03482493  3.62867401 2.876799e-04
black        -0.293767980 0.05297556 -5.54534899 3.081966e-08
othrace      -0.472445058 0.19839750 -2.38130547 1.728801e-02
factor(dep)1  0.029866899 0.05414025  0.55165799 5.812070e-01
factor(dep)2  0.096186517 0.04686228  2.05253621 4.016868e-02
q2            0.073678072 0.15682593  0.46980799 6.385124e-01
q3           -0.038506537 0.15647802 -0.24608272 8.056281e-01
q4           -0.054949195 0.15656019 -0.35097809 7.256193e-01
q5           -0.144177912 0.15588797 -0.92488160 3.550714e-01
q6            0.003361318 0.16645573  0.02019347 9.838898e-01
agelt35      -0.162772168 0.03696505 -4.40340780 1.087509e-05
agegt54       0.229666708 0.05917173  3.88135887 1.051906e-04
durable       0.126557359 0.04814585  2.62862426 8.598901e-03
lusd         -0.175352572 0.04098250 -4.27871860 1.914271e-05
husd         -0.105224727 0.04490864 -2.34308443 1.916341e-02

In [71]:
library(xtable)
table<- matrix(0, 4, 4)
table[1,1]<-  s[1,1]
table[1,2]<-  s[1,2]
table[1,3]<-  boot1[[1]][1]
table[1,4]<-  "0.14902630"

table[2,1]<-  s[2,1]
table[2,2]<-  s[2,2]
table[2,3]<-  boot1[[1]][2]
table[2,4]<-  "0.03607677"

table[3,1]<-  s[3,1]
table[3,2]<-  s[3,2]
table[3,3]<-  boot1[[1]][3]
table[3,4]<-  "0.03462895"

table[4,1]<-  s[4,1]
table[4,2]<-  s[4,2]
table[4,3]<-  boot1[[1]][4]
table[4,4]<-  "0.06072504"
colnames(table)<- c("estimate","standard error","estimate -boot", "standard error-boot")
rownames(table)<- c("intercept", "T4TRUE", "female", "black")
table

estimate            standard error     estimate -boot     
intercept 2.17846232630815    0.159015067145698  2.17846232630815   
T4TRUE    -0.0716924841364635 0.0354632588775146 -0.0716924841364635
female    0.126368328293608   0.0348249327707101 0.126368328293608  
black     -0.293767980365652  0.0529755621788769 -0.293767980365652 
          standard error-boot
intercept 0.14902630         
T4TRUE    0.03607677         
female    0.03462895         
black     0.06072504

- $T4$: In the fourth treated group, says that is $7.2\%$ less likely to be unemployed for one more week compared with groups who dont belong to this fourth group. Regarding its standar error, according to ols method, the standar error is $0.0354632588775146$, meanwhile using bootstraping method the standar error is $0.03607677$.

- $female$: Aditionally, women are $12.6\%$ more likely to be unemployed for one more week compared with men. Regarding its standar error, according to ols method, the standar error is $0.0348249327707101$, meanwhile using bootstraping method the standar error is $0.03462895$.

- $black$: Finally, afro american people is $29.37\%$ less likely to be unemployed for one more week compared with hispanic or white people. Regarding its standar error, according to ols method, the standar error is $0.0529755621788769$, meanwhile using bootstraping method the standar error is $0.06072504$.

In this exercise, we included many control variables, but we only are interested at the standar errors of three of them (T4, female and black). In this summary table, we can notice that the estimaded coefficients are the same for the ols regression and bootstrap method, however, but the standard errors estimated with bootstraping are slightly larger than those estimated by ols, except for female variable.  

## Example 2

In this second example, $log(inuidur1)~T4+ (female+black)$ formula for estimation, so bootstraping is goíng to estimate 4 variables (intercept included).

In [38]:
boot.fn2 <- function(Penn, index)
  coef(lm(log(inuidur1)~T4+ (female+black), data = Penn, subset = index))

In [39]:
boot2 = boot(Penn, boot.fn2, 1000)
boot2


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Penn, statistic = boot.fn2, R = 1000)


Bootstrap Statistics :
       original        bias    std. error
t1*  2.04405651 -0.0005382010  0.02511209
t2* -0.08454696  0.0004711029  0.03447406
t3*  0.11403808 -0.0022149042  0.03373497
t4* -0.27551947 -0.0001436283  0.05892468

In [42]:
g = summary(
    lm(log(inuidur1)~T4+ (female+black)), data = Penn
    )$coef
g

Estimate    Std. Error t value   Pr(>|t|)    
(Intercept)  2.04405651 0.02586004 79.043051 0.000000e+00
T4TRUE      -0.08454696 0.03571086 -2.367542 1.794383e-02
female       0.11403808 0.03453344  3.302250 9.657341e-04
black       -0.27551947 0.05177848 -5.321120 1.075462e-07

In [67]:
library(xtable)
table<- matrix(0, 4, 4)
table[1,1]<-  g[1,1]
table[1,2]<-  g[1,2]
table[1,3]<-  boot2[[1]][1]
table[1,4]<-  "0.02511209"

table[2,1]<-  g[2,1]
table[2,2]<-  g[2,2]
table[2,3]<-  boot2[[1]][2]
table[2,4]<-  "0.03447406"

table[3,1]<-  g[3,1]
table[3,2]<-  g[3,2]
table[3,3]<-  boot2[[1]][3]
table[3,4]<-  "0.03373497"

table[4,1]<-  g[4,1]
table[4,2]<-  g[4,2]
table[4,3]<-  boot2[[1]][4]
table[4,4]<-  "0.05892468"
colnames(table)<- c("estimate","standard error","estimate -boot", "standard error-boot")
rownames(table)<- c("intercept", "T4TRUE", "female", "black")
table

estimate            standard error     estimate -boot     
intercept 2.04405650652255    0.0258600406523543 2.04405650652255   
T4TRUE    -0.0845469633415028 0.0357108591415169 -0.0845469633415028
female    0.11403808076688    0.0345334443288449 0.11403808076688   
black     -0.275519465097314  0.0517784751197931 -0.275519465097314 
          standard error-boot
intercept 0.02511209         
T4TRUE    0.03447406         
female    0.03373497         
black     0.05892468

- $T4$: In the fourth treated group, says that is $8.5\%$ less likely to be unemployed for one more week compared with groups who dont belong to this fourth group. Regarding its standar error, according to ols method, the standar error is $0.0357108591415169$, meanwhile using bootstraping method the standar error is $0.03447406$.

- $female$: Aditionally, women are $11.4\%$ more likely to be unemployed for one more week compared with men. Regarding its standar error, according to ols method, the standar error is $0.0345334443288449$, meanwhile using bootstraping method the standar error is $0.03373497$.

- $black$: Finally, afro american people is $27.5\%$ less likely to be unemployed for one more week compared with hispanic or white people. Regarding its standar error, according to ols method, the standar error is $0.0517784751197931$, meanwhile using bootstraping method the standar error is $0.05892468$.

In this exercise, we oncly included three variable (T4, female and black). In this summary table, we can notice that the estimaded coefficients are the same for the ols regression and bootstrap method, however, but the standard errors estimated with bootstraping are slightly larger than those estimated by ols, except for the T4 variable.

#### Additional comment: 

- There are differences between the estimated coefficients and standar errors examples 1 and 2 because, in the first example we included many control variables and we didn't include them inthe second example. For this reason, the estimated coefficients of the first example are lower.

## Causal Tree

#### Explain why we need to partitionate the data in three sets.

The division into three sub-sets of data has the purpose of obtaining more efficient results and falling into overfiting. We separate the data into three: split, estimation and test. First, in the split we find the ideal tree, second, in the ideal tree I look for the heterogeneous effects of the treatment, third, in the test group estimate the heterogenic effects to check if the true outcomes are equal to the estimated ones.

#### Why do we need to use the honest.causalTree function?

Unlike adaptive trees, honest trees divide the data in such a way that one part trains the data and the other tests. The use of honest trees helps not to fall into the problem of overfitting the outcomes. A honest estimation results with tree structure built on training sample (including cross validation) and leaf treatment and ect estimates taken from estimation sample. In this sense, we use the honest.causalTree function to use the split and estimation groups, taking into account the treatment. In this way, with the split data we can build the tree and with estimate data we can calculate the heterogeneous effects.

#### Explain in detail the creation of the tree and how do you choose the optimal pruned tree
The data must be divided into three exactly equal parts. This division are called "divided", "estimated" and "tested". The “divided part” is used to create the tree that is not pruned yet and works with the honest causal tree. Let’s remember that there is also another way which is the "adaptive" (which executes the efficient way of how the trees should be divided and with that same splitted data we execute to get the heterogeneous effects of data). What this causal forest method does is to divide, estimate and test with three diferent parts. Remember provide the formula. The estimation data is used once the tree is created and we estimate the heterogeneous effect on each leaf. We want at the end to have at least one case on the last leaf.
To select the optimal pruned tree we use cross validation as the estimation type and the option of how the tree is going to be opened through "CT". We will also have a parameter equal to zero because we will not yet prune and, finally, we introduce the commands that validate honest division and cross validation (split.Honest = TRUE, cv.Honest = True).